### PR TITLE
feat(useWorkerLoad): add new hook

### LINKED ISSUE #11: [Feature] useStartWorker - A worker that starts immediately
# Motivation

We're often using a pattern where we create a worker and immediately call its callback, in order to load data when a component mounts. Also, we regularly set states within the worker's effect since we need to save its result.

# Proposal

`useStartWorker` is a hook that receives an async function that is called immediately and stores its result in a state. Its syntax is almost identical to `useWorker`, but instead of returning the `callback` attribute, it returns a `data` state that contains the function's return.

```ts
const { error, isLoading, data } = useStartWorker(async () => {
  const result = await loadDataFromAPI(arg);
  // manipulate result
  return result;
});
```

It may be necessary to include some way of providing the state's initial value if `data` can't be `undefined`:

```ts
const { error, isLoading, data } = useStartWorker(async () => {
  const { id } = await loadDataFromAPI(arg);
  return { id };
}, { id: -1 });
```

I can't think of any cases where this hook would require dependencies, since it should only be called when the component mounts. Consequently, its function shouldn't receive any arguments (other than the initial value, perhaps).

Better names are welcome, since `useStartWorker` does not sound terribly clever 🙃

### PR DESCRIPTION
**Pull request by fjcaetano**
fix #11

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -179,3 +179,46 @@ export const useLazyRef = <T extends any>(
...
+/**
+ * Starts loading a worker immediately and handle loading, error and result states.
+ *
+ * See [[useWorker]] for more details
+ *
+ * @param worker An asynchronous function that returns data, which is saved into a state
+ * @param initialValue The data's initial value. Optional when Data may be undefined
+ */
+export const useWorkerLoad = <Data extends any>(
+  worker: () => Promise<Data>,
```
Comment (uqmessias) [inline src/index.tsx:527]: What happens if the worker is a function that depends on other data? Shouldn't we have a `dependencies?: Dependencies` here?

Comment (uqmessias) [reply src/index.tsx:527]: If the idea is to not allow the `worker` to be updated, shouldn't it be explained some how in the JSDoc?

File: src/index.tsx [hunk 2]
```diff
@@ -179,3 +179,46 @@ export const useLazyRef = <T extends any>(
...
+    error,
+    setError,
+    setIsLoading,
+    callback,
+  } = useWorker(async () => {
+    setData(await worker());
+  }, []);
+
+  // start loading immediately
+  useDidMount(callback);
```
Comment (uqmessias) [inline src/index.tsx:547]: 💯

File: src/index.tsx (hunk 1 above)
Comment (fjcaetano) [reply src/index.tsx:527]: The idea is that the `worker` function is only executed when the component gets mounted, so there's no need to update its dependencies. That's also why the callback is not exposed.

I added the `retry` thinking of a scenario with errors, but I don't know if it should exist when `error` is `undefined`. What do you think? The way that it is now, maybe `reload` or `rework` would be better names for it. If we decide to allow the user to call `retry` whenever they want, then we may need to add `worker` as a dependency of `useWorker`, for the case where the function is updated with new values.

Comment (danibonilha) [reply src/index.tsx:527]: If the idea is to retry only on error, maybe creating a boolean `retryOnError` something like that and handling inside `useWorkerLoad` would avoid requiring dependencies.

Comment (fjcaetano) [reply src/index.tsx:527]: But when would `retry` happen in that case? Immediately? I don't know if that's the best case. How would we prevent infinite retries? IMO, it'd be better if the user triggered a new retry

Comment (fjcaetano) [reply src/index.tsx:527]: Idea: what if `error` was typed in the return as: `error?: { value: Error; retry: () => void }`. With the coalescing syntax, it'd be simple to access `error?.value` and `error?.retry()`

Comment (danibonilha) [reply src/index.tsx:527]: 🤔 I like this idea

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,21 @@
+# [0.3.0-alpha1](https://github.com/fjcaetano/react-hook-utilities/compare/v0.2.1...v0.3.0-alpha1) (2019-11-11)
```
Comment (uqmessias) [inline CHANGELOG.md]: 💯

**Review by uqmessias — APPROVED** [review]
Good stuff

**Comment by fjcaetano** [comment]
**TODO**:
- [x] Docs for this hook are not being exported with distribution code

**Pull request merged**